File: functions/scope.js
// javascript has function scope, and it hoists variables and functions
// so it is best to define all the variables at the top of the function.

var foo = function () {
  var a = 3, b = 5;

  var bar = function () {
    var c = 11;

    b = 7;

    console.log("inside bar", a, b, c);

    a += b + c;

    console.log("inside bar", a, b, c);
  }

  console.log("before bar", a, b);

  bar();

  console.log("after bar", a, b);
}

foo();

File: functions/scope.rb
foo = -> {
  a = 3
  b = 5

  bar = -> {
    b = 7
    c = 11

    puts "inside bar: #{a}, #{b}, #{c}"

    a += b + c

    puts "inside bar: #{a}, #{b}, #{c}"
  }

  puts "before bar: #{a}, #{b}"

  bar.()

  puts "after bar: #{a}, #{b}"
}

foo.()
